### ISSUE TITLE
Analyzer suggestion: unnecessary use of `unsafe`

### ISSUE THREAD

**Issue by stephentoub** [dotnet/roslyn#48031]
**Brief description:**

We frequently find ourselves initially writing some method to use unsafe code / pointers, and so annotate the method as `unsafe`.  Later, generally as `Span<T>` is employed more or as new framework APIs are utilized to improve the method, the method no longer uses pointers, but we sometimes forget to remove the `unsafe` keyword.  It would be nice-to-have if there were an analyzer we could enable to flag such uses of `unsafe` that could be removed.

**Languages applicable:**

C#

**Code example that the analyzer should report:**

```C#
private unsafe void ToUpper(Span<char> span)
{
    for (int i = 0; i < span.Length; i++)
        span[i] = char.ToUpperInvariant(span[i]);
}
```

**Comment by sharwell** [comment]
I've wanted this for sure

**Comment by Youssef1313** [comment]
I'm interested in implementing this if it's approved in the design meeting.

**Comment by CyrusNajmabadi** [comment]
I'm willing to pre-emptively approve this.  We would take this.  Open question is on if it's feasible to encode the rules easily enough to put this in teh IDE.  My preference would be more that this be done in the compiler with a hidden diagnostic.

**Comment by vatsalyaagrawal** [comment]
Design meeting notes: This should come from Compiler, request hidden diagnostic from the compiler.
